### ISSUE TITLE
Full Library Sync only gets first page

### ISSUE THREAD

**Issue by girino** [openaudible/openaudible#191]
This seems to be the same problem as in issue #185, which is closed. I have 500+ books, but only first 20 were listed. When i change "titles per page" to 50, i can get 50 more. Changing the sort order gives me still more 50. In the end, by tweeking with sorting and filtering, i got 151 titles. Still way too low compared to the 500+.

I'm running OSX high sierra with 1.5.3 binary release.

**Comment by openaudible** [comment]
Full Sync is supposed to start at page 1 and go until it can't find a "Next button". Quick does the same thing, but stops as soon as no new books are found on a page. 

We've had reports where it fails. If you wanted to submit a bug, you could restart OA, connect, do a full sync, and if the problem happens, submit a bug report from the Tools menu and reference issue #191.

**Comment by sonnenlanze** [comment]
same here. Library Sync only gives back page one (of more than 10)

> Okt 03, 2019 10:55:29 AM org.openaudible.desktop.swt.manager.menu.CommandCenter execute
> INFORMATION: Command: Rescan_Library
> Okt 03, 2019 10:55:29 AM org.openaudible.audible.AudibleScraper _fetchLibrary
> INFORMATION: ******************* LIBRARY PAGE 1
> Okt 03, 2019 10:55:29 AM org.openaudible.audible.AudibleScraper setPageFilter
> WARNUNG: setPageFilter Unimplemented
> Okt 03, 2019 10:55:29 AM org.openaudible.audible.parser.LibraryParser parseLibraryFragment
> INFORMATION: Library page contains: 20 book(s)
> Okt 03, 2019 10:55:29 AM org.openaudible.audible.AudibleScraper _fetchLibrary
> INFORMATION: pageNum=1 total=20 new=20 partial=0
> Parsed:20 books
> BK_AREN_000439DE: 	new:The Fires of Heaven: Book Five o
> BK_AREN_000421DE: 	new:The Shadow Rising: Book Four of 
> BK_AREN_000420DE: 	new:The Dragon Reborn: Book Three in
> BK_AREN_000371DE: 	new:The Great Hunt: The Wheel Of Tim
> BK_ADBL_050938DE: 	new:Unfettered III: New Tales by Mas
> BK_AREN_000370DE: 	new:The Eye of the World: The Wheel 
> BK_RHUK_002292DE: 	new:The Magicians, Book 1
> BK_ADBL_027985DE: 	new:Skyward
> BK_ADBL_020668DE: 	new:Unfettered II: New Tales by Mast
> BK_HCUK_003091DE: 	new:The Core: The Demon Cycle, Book 
> BK_HCUK_002005DE: 	new:The Skull Throne: The Demon Cycl
> BK_HCUK_001861DE: 	new:The Daylight War: The Demon Cycl
> BK_HCUK_001860DE: 	new:The Desert Spear: The Demon Cycl
> BK_HCUK_001859DE: 	new:The Painted Man: The Demon Cycle
> BK_ORIO_001018DE: 	new:Arcanum Unbounded: The Cosmere C
> BK_ORIO_000895DE: 	new:The Bands of Mourning: A Mistbor
> BK_ORIO_000867DE: 	new:Shadows of Self: A Mistborn Nove
> BK_PENG_003900DE: 	new:Brief Cases
> BK_PENG_002891DE: 	new:Shadowed Souls
> BK_TWUK_000709DE: 	new:Side Jobs: Stories from the Dres
> 
> Okt 03, 2019 10:55:29 AM org.openaudible.Audible updateLibrary
> INFORMATION: Updated 20 books
> Okt 03, 2019 10:55:29 AM org.openaudible.util.HTMLUtil writeFile
> INFORMATION: wrote 66069 to disk:C:\Users\Sunny\OpenAudible\tmp\saved_books.json
> Okt 03, 2019 10:55:29 AM org.openaudible.desktop.swt.manager.AudibleGUI downloadAndConvertWithDialog
> INFORMATION: You have 20 book(s) to download.
>  autoConvert=false

**Comment by openaudible** [comment]
I will add more logging info. Make sure your books are sorted newest to oldest. Its probably something not quite the same as everyone else.

**Comment by rockandrollnerd** [comment]
First time I used it right after install it only pulled up 20 books too and neither quick or full sync did anything. 
After trying a bunch of stuff here's what worked for me on Windows 10:
Open browser in openaudible (CTL+B), go to audible.com and if you're logged in - log out.
Close browser. Close openaudible.
Re-launch openaudible and connect to audible.com again (CTL+K).
Scroll down to the bottom where it says: Show 20 titles per page and change that to 50.
Openaudible automatically pulled the other 30 books.
Went to 2nd page and it pulled the rest (one more book in my case bc I only have 51 in there)
Hope this works for the rest of you.

**Comment by openaudible** [comment]
I added new code in 1.5.5 to help debug this.. Run again and look for text that says: 
no 'next' page found... exiting scan. Debug=
and include the Debug= message..

If you you see that, you can send a bug report from the Tools menu and I'll get a copy to look at.

**Comment by m-j-r** [comment]
I'm getting the same issue - here is my Debug= output:

1 buttons
libPageFile=/home/mjr/OpenAudible/tmp/lib_1.html
Oct 23, 2019 3:42:10 PM org.openaudible.Audible updateLibrary
INFO: Updated 20 books
Oct 23, 2019 3:42:10 PM org.openaudible.util.HTMLUtil writeFile
INFO: wrote 46266 to disk:/home/mjr/OpenAudible/tmp/saved_books.json
Oct 23, 2019 3:42:10 PM org.openaudible.Audible save
INFO: Save books took 2 ms. to complete.
Oct 23, 2019 3:42:10 PM org.openaudible.desktop.swt.manager.AudibleGUI downloadAndConvertWithDialog
INFO: You have 19 book(s) to download.
 autoConvert=true

**Comment by openaudible** [comment]
One button looks like a problem. That means on your entire library page has only one clickable link. What does the OA browser show (CMD-B) ? You should be able to watch the browser to go to each page of your library. Also, you could open the /home/mjr/OpenAudible/tmp/lib_1.html file in a browser and see if it looks like a library page, or some other kind of "Log into audible/not a robot" page.

**Comment by m-j-r** [comment]
I watched the browser go into the library and also accessed the file in my ${HOME}/OpenAudible and they both look like the first page of my library.

**Comment by alvarow** [comment]
This also happens to me, but if I manually click next on the little OA browser page, OA refreshes itself after a second of two and the new books get added to the list of books. What's odd is that some books show up in aax/ folder twice, one with .1 extension and the other without.

**Comment by openaudible** [comment]
I added some debugging code. Feel free to submit a new bug report. Open version 1.6 (or latest), connect, do a scan, and if it only gets page one, send a bug report and reference #191.

**Comment by mark0978** [comment]
Going to make a guess at what is going on here from browsing the source and looking at the page.  The next button is found using this code which is looking for a `<button>`
```
public HtmlElement getNextPage(HtmlPage page) {
		HtmlElement next = null;
		DomNodeList<HtmlElement> buttons = page.getDocumentElement().getElementsByTagName("button");
		
		for (HtmlElement a : buttons) {
			// System.out.println(a);
			if (a.toString().toLowerCase().contains("pagenext")) {
				assert (next == null);
				next = a;
			}
		}
		
		return next;
	}
```
Which some designer has replaced with a different version on the page.  That version looks like this:
```
<li class="bc-list-item">
    <span class="bc-button  bc-button-secondary  nextButton refinementFormButton  bc-button-small bc-button-inline">
      <a class="bc-button-text" data-url="/lib?pf_rd_p=b713f373-2416-445c-8747-dee82ab9d12e&amp;pf_rd_r=87V243GAB8KNYBD7JSYP&amp;ref=a_lib_c5_pageNext" data-name="page" data-value="2" href="/lib?pf_rd_p=880a7626-bffe-4598-ad77-4ed67dc2e0f6&amp;pf_rd_r=DWKNMNE5JR8RV0HMSP07&amp;ref=a_howtolist_t1_navTop_pl0&amp;page=2" tabindex="0">
        <span class="bc-text bc-button-text-inner bc-size-small">
           <i aria-hidden="true" class="bc-icon bc-icon-chevron-right"></i>
           <span class="bc-text bc-pub-offscreen">Go forward a page</span>
        </span>
      </a>
    </span>
</li>
```
And if you go after it looking for `.nextButton` (a class) that should give you the `<span>` that contains the `<a>` and gives you what you need to move forward.

All that said, I don't have all the tooling setup for maven and Java et all...  so that's about all the help I can give you.

**Comment by reidar-cederqvist** [comment]
you could add "&page=<num>" to the URL where <num> is increasing until you get an empty page.

**Comment by openaudible** [comment]
Thanks for the reports... I think I got it with version 1.6.1. Let me know!

**Comment by AshbyGeek** [comment]
I had already gotten my full library by manually pressing the next button, but I just tried your latest version anyway. It was flipping through the pages just fine but threw a "java.lang.NullPointerException" after about the 4th page.

**Comment by openaudible** [comment]
Thanks @mark0978, you were correct.. was looking for "button" but that changed to "a" anchors for some browsers.. @AshbyGeek thanks for the report... that will be fixed in 1.6.1.
Will close and please re-open or create new issue if anything comes up.

**Comment by AshbyGeek** [comment]
My comment was made after installing 1.6.1. I happened to be online reading through this issue when you released that version so I installed it and gave it a try. 

Here's a screenshot for what it's worth. Looks like it might have croaked on the empty last page?
![image](https://user-images.githubusercontent.com/83048/68911009-a8786880-0721-11ea-8640-f8af37d7d4e5.png)
